src/server/raft_client.rs

```diff
@@ -69,6 +70,25 @@ static CONN_ID: AtomicI32 = AtomicI32::new(0);
 
 const _ON_RESOLVE_FP: &str = "transport_snapshot_on_resolve";
 
+pub(crate) fn observe_read_index_duration(h: &Histogram, b: &BatchRaftMessage, t: MessageType) {
+    let mut now = Duration::ZERO;
+    for msg in b.get_msgs() {
+        let m = msg.get_message();
+        let typ = m.get_msg_type();
+        let ctx = m.get_context();
+        if typ == t && ctx.len() > 0 {
+            if let Some(ts) = ReadIndexContext::get_ts_from_id(ctx) {
+                if now.is_zero() {
+                    now = SystemTime::UNIX_EPOCH
+                        .elapsed()
+                        .expect("clock may have gone backwards");
+                }
+                h.observe(now.saturating_sub(ts).as_secs_f64());
```

Comment:
This is cross machine time difference, which is not accurate due to time difference on machines. 
The duration calculated here can be used as a reference, but could be misleading too

Reply:
The function is called when a store: 1) sends heartbeat requests; 2) receives heartbeat responses , the timespecs is obtained on the same store, it's not cross machine time difference. However, the clock might be updated by NTP periodically, so you are right, the duration might be inaccurate, but not too much (IMO).

This is a debug patch for investigating slow read index issue, such as https://clinic.pingcap.com/portal/#/orgs/1372813089196927732/clusters/7211629990340078758?from=1701723600&to=1701724800 . It does not directly tell us why read index becomes slow, but provide us more information. I'll turn this PR to draft.